tracing-subscriber/src/fmt/fmt_layer.rs

```diff
@@ -0,0 +1,532 @@
+    N: for<'writer> FormatFields<'writer> + 'static,
+    E: FormatEvent<S, N> + 'static,
+    W: MakeWriter + 'static,
+{
+    #[inline]
+    fn make_ctx<'a>(&'a self, ctx: Context<'a, S>) -> FmtContext<'a, S, N> {
+        FmtContext {
+            ctx,
+            fmt_fields: &self.fmt_fields,
+        }
+    }
+}
+
+/// A formatted representation of a span's fields stored in its [extensions].
+///
+/// Because `FormattedFields` is generic over the type of the formatter
+/// that produced it, multiple versions of a span's formatted fields can be
+/// stored in the [`Extensions`][extensions] type-map. This means that when
+/// multiple formatters are in use, each can store its own formatted
+/// representation without conflicting.
+///
+/// [extensions]: ../registry/extensions/index.html
+#[derive(Default)]
+pub struct FormattedFields<E> {
+    _format_event: PhantomData<fn(E)>,
+    /// The formatted fields of a span.
+    fields: String,
+}
+
+impl<E> fmt::Debug for FormattedFields<E> {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        f.debug_struct("FormattedFields")
+            .field("fields", &self.fields)
+            .finish()
+    }
+}
+
+impl<E> fmt::Display for FormattedFields<E> {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        write!(f, "{}", self.fields)
+    }
+}
+
+impl<E> Deref for FormattedFields<E> {
+    type Target = String;
+    fn deref(&self) -> &Self::Target {
+        &self.fields
+    }
+}
+
+// === impl FmtLayer ===
+
+impl<S, N, E, W> layer::Layer<S> for Layer<S, N, E, W>
+where
+    S: Subscriber + for<'a> LookupSpan<'a>,
+    N: for<'writer> FormatFields<'writer> + 'static,
+    E: FormatEvent<S, N> + 'static,
+    W: MakeWriter + 'static,
+{
+    fn new_span(&self, attrs: &Attributes<'_>, id: &Id, ctx: Context<'_, S>) {
+        let span = ctx.span(id).expect("Span not found, this is a bug");
+        let mut extensions = span.extensions_mut();
+
+        let mut buf = String::new();
+        if self.fmt_fields.format_fields(&mut buf, attrs).is_ok() {
+            let fmt_fields = FormattedFields {
+                fields: buf,
+                _format_event: PhantomData::<fn(N)>,
+            };
+            extensions.insert(fmt_fields);
+        }
+    }
+
+    fn on_record(&self, id: &Id, values: &Record<'_>, ctx: Context<'_, S>) {
+        let span = ctx.span(id).expect("Span not found, this is a bug");
+        let mut extensions = span.extensions_mut();
+
+        let mut buf = String::new();
+        if self.fmt_fields.format_fields(&mut buf, values).is_ok() {
+            let buf = match extensions.get_mut::<FormattedFields<Self>>() {
+                Some(fields) => format!("{}{}", fields.fields, buf),
+                None => buf,
+            };
+            let fmt_fields = FormattedFields {
+                fields: buf,
+                _format_event: PhantomData::<fn(N)>,
+            };
+            extensions.insert(fmt_fields);
+        }
```

Comment:
from a perf perspective, the behavior when there's already a formatted representation for a span in the extensions is really not great. we appear to be creating a new `String`, writing the new fields to it (an allocation), using `format!` to append the previous fields to the new ones, returning _another_ new `String`, and then putting the `String` returned by `format!` back into the extensions. the previous value in extensions will then be dropped, along with the buffer the fields were formatted into, causing two _de_allocations.

i think it would be better if we checked if there were already formatted fields for the span, and if so, called `format_fields` with a mutable reference to the _existing_ string, instead. that may still allocate, if there isn't enough capacity in the buffer, but it would allocate _once_ in the worst case (and no allocations if the buffer has sufficient capacity remaining). also, it wouldn't result in always deallocating two additional strings.

if you'd prefer to change this in a follow-up, that's okay, but i think it really ought to be addressed. this method doesn't get called _that_ much in practice, but it would proably be _simpler_ if we changed the behavior here.

Reply:
@davidbarsky something like
```rust
    fn on_record(&self, id: &Id, values: &Record<'_>, ctx: Context<'_, S>) {
        let span = ctx.span(id).expect("Span not found, this is a bug");
        let mut extensions = span.extensions_mut();
        if let Some(FormattedFields { ref mut fields, .. }) = extensions.get_mut::<FormattedFields<Self>>() {
            let _ = self.fmt_fields.format_fields(fields, values);
        } else {
            let mut buf = String::new();
            if self.fmt_fields.format_fields(&mut buf, values).is_ok() {
                let fmt_fields = FormattedFields {
                    fields: buf,
                    _format_event: PhantomData::<fn(N)>,
                };
                extensions.insert(fmt_fields);
            }
        }
    }
```
should work

Reply:
it did, thank you!